core/src/main/java/io/grpc/LoadBalancer.java

```diff
@@ -169,8 +170,75 @@ public abstract void handleSubchannelState(
      *
      * @param affinity the affinity attributes provided via {@link CallOptions#withAffinity}
      * @param headers the headers container of the RPC. It can be mutated within this method.
+     * @deprecated use {@link #pickSubchannel(PickSubchannelArgs)}
      */
-    public abstract PickResult pickSubchannel(Attributes affinity, Metadata headers);
+    @Deprecated
+    public PickResult pickSubchannel(Attributes affinity, Metadata headers) {
+      throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Make a balancing decision for a new RPC.
+     *
+     * @param args the pick arguments
+     */
+    // TODO(lukaszx0) make it abstract once deprecated overload will be removed.
+    public PickResult pickSubchannel(PickSubchannelArgs args) {
+      return pickSubchannel(args.getCallOptions().getAffinity(), args.getHeaders());
+    }
+  }
+
+  /**
+   * Provides arguments for a {@link SubchannelPicker#pickSubchannel(PickSubchannelArgs)}.
+   */
+  public abstract static class PickSubchannelArgs {
+
+    /**
+     * Call options.
+     */
+    public abstract CallOptions getCallOptions();
+
+    /**
+     * Call metadata.
+     */
+    public abstract Metadata getHeaders();
+
+    /**
+     * Call method.
+     */
+    public abstract MethodDescriptor<?, ?> getMethodDescriptor();
+
+    /**
+     * Returns true if the given object is also a {@link PickSubchannelArgs} with an equal
+     * getter values.
+     */
+    @Override
+    public boolean equals(Object o) {
```

Comment:
Actually, let's not have equals/hashCode. If you're comparing them, Your Doing It Wrong™. That also prevents having the problem later where it is hard to figure out what to consider "equal".